Step5-Refactoring

goodbyeyo opened the pull request:
남재님 안녕하세요

Step5 리팩토링 미션 리뷰 부탁드립니다.

Step4 PR 피드백 중에서 ```가짜 생성자를 통해서 car로 score을 생성하는 구조는 어떨까요?```
이부분에 대한 피드백 반영이 제대로 되었는지
그리고 가짜 생성자가 부생성자와 같은 의미로 사용되는지도 궁금합니다. 

마지막으로 구현한 코드 중 개선이 좀 더 필요한 부분이나 
다른 방법으로 접근 또는 구조를 개선 할 수 있는 내용이 있다면 알려주시면 감사하겠습니다!

goodbyeyo committed:
refactor: Cars class forEach & repeat 문-> map 키워드 사용하도록 변경

goodbyeyo committed:
refactor: Step4 피드백 반영

goodbyeyo committed:
refactor:
 1. controller 패키지 추가
 2. RacingCarGame class 이동

goodbyeyo committed:
refactor: Winners 필드명 수정, 구조 변경

namjackson commented on src/main/kotlin/racingcar/domain/Score.kt:
```diff
@@ -1,12 +1,13 @@
-
-    val point: Int = car.point
-    val carName: String = car.carName.name
+    constructor(car: Car) : this(car.point, car.carName.name)
+    constructor(point: Int) : this(point, BLANK_NAME)
```
> Step4 PR 피드백 중에서 가짜 생성자를 통해서 car로 score을 생성하는 구조는 어떨까요?
> 이부분에 대한 피드백 반영이 제대로 되었는지
> 그리고 가짜 생성자가 부생성자와 같은 의미로 사용되는지도 궁금합니다.

가짜 생성자와 부생성자는 비슷한 역할을 하긴하지만,
각각 다른 특징을 가지고 있어요 :) 
장단점이 존재하니 가짜생성자에 대해서도 읽어보시면 좋을거같아요!
https://jaeyeong951.medium.com/fake-contructor-7aef39f98a25

저도 개인적으로는 부생성자를 활용하기보다는 디폴트값이나, 가짜생성자를 주로 활용하는거같아요!

namjackson commented on src/main/kotlin/racingcar/domain/Car.kt:
```diff
@@ -4,11 +4,9 @@ import racingcar.strategy.NumberGeneratorStrategy
 
 class Car(
     point: Int = 0,
-    name: CarName,
+    name: CarName = CarName(BLANK_NAME),
```
추가로 고민해볼법한 포인트들은 
Car를 불변객체로 설계하는 부분이 있을거 같아요! 
이부분은 사실 다음 미션 ( 로또, 블랙잭 등 ) 에서 주로 다루니
다음미션에서 진행해보셔도 좋을거같아요!

namjackson commented on src/main/kotlin/racingcar/domain/Winners.kt:
```diff
@@ -1,13 +1,12 @@
 package racingcar.domain
 
 class Winners(carNames: List<CarName>) {
```
사실 Winners의 names도 가변컬렉션으로 유지할필요가 있을까요?
addWinner가 아닌, winner names를 주입받는 구조라면 좀더 구조상 안전할거 같단생각이들어요

namjackson commented on src/main/kotlin/racingcar/domain/Cars.kt:
```diff
@@ -16,10 +16,8 @@ class Cars(carNames: CarNames, carCount: PositiveNumber) {
-        repeat(carNames.count()) {
-            val car = Car(carNames.name(it))
-            _cars.add(car)
-        }
+        val cars = carNames.map { Car(it) }
+        _cars.addAll(cars)
         return _cars
     }
```
`private val cars: List<Car> =  carNames.map { Car(it) }`
불필요한 코드들을 많이 제거할수 있을거같아요 :)
map 같은 다양한 컬렉션함수, 체이닝 함수들이 존재하니까 이부분 사용하시는데 익숙해지시면 좋을거같아요!

namjackson commented on src/main/kotlin/racingcar/domain/Scores.kt:
```diff
@@ -26,19 +26,12 @@ class Scores(scores: List<Score>) {
-                winners.addWinner(score.carName)
-            }
-        }
+        _scores.filter { score -> score.isSameScore(maxScore) }
+            .forEach { winners.addWinner(it.carName) }
```
```suggestion
       val  winners =  _scores.filter { score -> score.isSameScore(maxScore) }
       return  Winners(winners)
```
이런 구조도 깔끔할거같아요 :)

namjackson reviewed (changes requested):
안녕하세요 상욱님 
자동차 미션 고생하셨어요! 
미션종료하기전 몇가지 추가적으로 코멘트 남겼으니, 확인해주세요 :)

namjackson commented:
아마 다음 요청시에는 크게 피드백 드릴부분이 없을거같아서,
미션이 마감될거같아요!
추가적으로 코드정리나, 수정하실 부분이 도전하시면 좋을거같아요! 
자동차 미션 고생하셨습니다!

goodbyeyo committed:
refactor: 피드백 반영
1. secondary constructor -> fake constructor 변경
2. mutable collection -> immutable collection 변경
3. 미사용 매서드 정리

namjackson commented on src/main/kotlin/racingcar/domain/Score.kt:
```diff
@@ -6,10 +6,15 @@ data class Score(
     }
 }
+
+fun Score(car: Car): Score {
+    return Score(car.point, car.carName.name)
+}
+
+fun Score(point: Int): Score {
+    return Score(point, BLANK_NAME)
+}
```
```
class Score private constructor(val ...
    companion object {
        fun create(car: Car): Score 
```
이런 방식으로도 많이 사용할수 있으니 참고만해주세요 :)

namjackson reviewed (approved):
안녕하세요 상욱님!
자동차미션 고생하셨습니다 :)
미션은 여기서 종료할게요, 그동안 고생하셨습니다! 
남은 미션들도 화이팅입니다!! :)
